compiler/rustc_query_system/src/dep_graph/graph.rs

```diff
@@ -385,16 +386,40 @@ impl<K: DepKind> DepGraph<K> {
     ) -> (R, DepNodeIndex)
     where
         OP: FnOnce() -> R,
+    {
+        self.with_hash_task(cx, dep_kind, op, |task_deps| {
+            // The dep node indices are hashed here instead of hashing the dep nodes of the
+            // dependencies. These indices may refer to different nodes per session, but this isn't
+            // a problem here because we that ensure the final dep node hash is per session only by
+            // combining it with the per session random number `anon_id_seed`. This hash only need
+            // to map the dependencies to a single value on a per session basis.
+            let mut hasher = StableHasher::new();
+            task_deps.reads.hash(&mut hasher);
```

Comment:
Idea for a simpler solution: could we hash the `DepNode` instead of the `DepNodeIndex` to avoid order-dependence? Could this solve the issue without having to make client code do the hashing?

Reply:
This is good advice, but `DepNode` in `TaskDeps` only exists in the case of `#[cfg(debug_assert)]`. I'm using `Hash` bound now in where clause of `with_task_hash`, which can avoid hashing in client code.